From 46951812bd6cd7947ac2efd2a85d8bdfa26cd11f Mon Sep 17 00:00:00 2001
From: Limeng <Meng.Li@windriver.com>
Date: Mon, 16 Jan 2017 14:47:24 +0800
Subject: [PATCH] driver: usb: add code to check acts_udc_controller pointer
 bofore using it
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

When enable usb aotg driver by kernel configure CONFIG_USB_AOTG
and cat /proc/acts_udc, kernel will crash and core trace as below:
Unable to handle kernel NULL pointer dereference at virtual address 0000032c
pgd = e4538000
[0000032c] *pgd=24522831, *pte=00000000, *ppte=00000000
Internal error: Oops: 17 [#1] PREEMPT SMP ARM
Modules linked in: iptable_filter ip_tables x_tables openvswitch libcrc32c
pvrsrvkm drm atc260x_irkeypad sch_fq_codel softdog nfsd
CPU: 1 PID: 530 Comm: cat Not tainted 4.1.21-WR8.0.0.0_standard #5
Hardware name: gs705a
task: e6b34ec0 ti: e451c000 task.ti: e451c000
PC is at aotg_output_current_ep_status+0x40/0x1fc
LR is at vprintk_emit+0x4fc/0x580
pc : [<c058a868>]    lr : [<c009775c>]    psr: 60070013
sp : e451de70  ip : e451ddc8  fp : e451deb4
r10: e451ded8  r9 : b6db7000  r8 : e451df80
r7 : 00000001  r6 : 00000000  r5 : 00000000  r4 : 00000000
r3 : c1732db4  r2 : e6b34ec0  r1 : c0beb5f2  r0 : 00000021
Flags: nZCv  IRQs on  FIQs on  Mode SVC_32  ISA ARM  Segment user
Control: 10c5387d  Table: 2453804a  DAC: 00000015
Process cat (pid: 530, stack limit = 0xe451c210)
Stack: (0xe451de70 to 0xe451e000)
de60:                                     00000001 00000000 00001000 c01ccb90
de80: 00000817 00001000 00020000 e9506600 00020000 00000000 00000001 e451df80
dea0: b6db7000 e451ded8 e451dec4 e451deb8 c058aa3c c058a834 e451df14 e451dec8
dec0: c01cd058 c058aa30 00020000 e95163c8 e9506630 e95163c0 00000000 00000000
dee0: e451df14 e451def0 c03d4b20 c01cce98 eaabacc0 00000001 00000000 c000f5c8
df00: e451c000 00000000 e451df34 e451df18 c0206528 c01ccea4 c02064a8 e95163c0
df20: e451df80 00020000 e451df4c e451df38 c01a7610 c02064b4 b6db7000 e95163c0
df40: e451df7c e451df50 c01a7d5c c01a75f4 c01c5bdc c01c5b38 e95163c0 e95163c0
df60: b6db7000 00020000 c000f5c8 e451c000 e451dfa4 e451df80 c01a854c c01a7cd4
df80: 00000000 00000000 00020000 00020000 b6db7000 00000003 00000000 e451dfa8
dfa0: c000f3c0 c01a8500 00020000 00020000 00000003 b6db7000 00020000 0002a2ac
dfc0: 00020000 00020000 b6db7000 00000003 7fffe000 00000000 b6f44000 00020000
dfe0: 00000000 beb86b34 000147b4 b6e98da0 60070010 00000003 00000000 00000000
[<c058a868>] (aotg_output_current_ep_status) from [<c058aa3c>] (acts_udc_proc_show+0x18/0x20)
[<c058aa3c>] (acts_udc_proc_show) from [<c01cd058>] (seq_read+0x1c0/0x41c)
[<c01cd058>] (seq_read) from [<c0206528>] (proc_reg_read+0x80/0x94)
[<c0206528>] (proc_reg_read) from [<c01a7610>] (__vfs_read+0x28/0x48)
[<c01a7610>] (__vfs_read) from [<c01a7d5c>] (vfs_read+0x94/0xfc)
[<c01a7d5c>] (vfs_read) from [<c01a854c>] (SyS_read+0x58/0x98)
[<c01a854c>] (SyS_read) from [<c000f3c0>] (ret_fast_syscall+0x0/0x60)
Code: e3a04090 e3550000 e5936000 e0246594 (e594832c)
This issue is introduced by commit ed785db7c3bc
(“usb: add owl usb2.0 device driver”) when merge SDK patch related
with usb driver.
Because when usb works as host mode, the pointer acts_udc_controller
is NULL, we need to check the pointer before using it.

Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/usb/aotg/aotg_udc_debug.c |    3 +++
 1 files changed, 3 insertions(+), 0 deletions(-)

diff --git a/drivers/usb/aotg/aotg_udc_debug.c b/drivers/usb/aotg/aotg_udc_debug.c
index 0ccb76a..8432801 100755
--- a/drivers/usb/aotg/aotg_udc_debug.c
+++ b/drivers/usb/aotg/aotg_udc_debug.c
@@ -117,6 +117,9 @@ void aotg_output_current_ep_status(struct seq_file *m)
 	struct aotg_udc *udc;
 	int i;
 
+	if(acts_udc_controller == NULL)
+		return;
+
 	printk("%s %d\n",__func__,__LINE__);
 	for (i = 0; i < AOTG_UDC_NUM_ENDPOINTS; i++) {
 		ep = &acts_udc_controller->ep[i];
-- 
1.7.5.4

